test/integration/start_tls_integration_test.cc

```diff
@@ -0,0 +1,533 @@
+/*
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.pb.h"
+#include "envoy/extensions/access_loggers/file/v3/file.pb.h"
+#include "envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.pb.h"
+
+#include "common/config/api_version.h"
+#include "common/network/utility.h"
+
+#include "extensions/filters/network/common/factory_base.h"
+#include "extensions/transport_sockets/tls/context_manager_impl.h"
+*/
+
+#include "envoy/network/filter.h"
+#include "envoy/server/filter_config.h"
+
+#include "extensions/filters/network/common/factory_base.h"
+#include "extensions/transport_sockets/raw_buffer/config.h"
+#include "envoy/config/transport_socket/raw_buffer/v2/raw_buffer.pb.h"
+#include "envoy/config/transport_socket/raw_buffer/v2/raw_buffer.pb.validate.h"
+#include "common/network/connection_impl.h"
+
+
+#include "test/integration/start_tls_integration_test.pb.h"
+#include "test/integration/start_tls_integration_test.pb.validate.h"
+#include "test/integration/integration.h"
+#include "test/integration/ssl_utility.h"
+//#include "test/integration/utility.h"
+#include "test/test_common/registry.h"
+#include "gtest/gtest.h"
+
+
+namespace Envoy {
+
+
+class StartTlsSwitchFilter : public Network::Filter {
+public:
+  // Network::ReadFilter
+  Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override;
+  Network::FilterStatus onNewConnection() override;
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
+    read_callbacks_ = &callbacks;
+  }
+
+  // Network::WriteFilter
+  Network::FilterStatus onWrite(Buffer::Instance& data, bool end_stream) override;
+  void initializeWriteFilterCallbacks(Network::WriteFilterCallbacks& callbacks) override {
+    write_callbacks_ = &callbacks;
+  }
+
+private:
+  Network::ReadFilterCallbacks* read_callbacks_{};
+  Network::WriteFilterCallbacks* write_callbacks_{};
+};
+
+// Network::ReadFilter
+Network::FilterStatus StartTlsSwitchFilter::onNewConnection() { return Network::FilterStatus::Continue; }
+
+Network::FilterStatus StartTlsSwitchFilter::onData(Buffer::Instance&, bool) {
+  printf("Data Received onData!\n");
+//  data.drain(data.length());
+  printf("%s\n", read_callbacks_->connection().transportProtocol().c_str());
+  if (read_callbacks_->connection().transportProtocol() ==
+      Extensions::TransportSockets::TransportProtocolNames::get().StartTls) {
+    // We run on top of the `STARTTLS` socket and can ask it to convert to SSL.
+  printf("Switching to TLS!\n");
+    read_callbacks_->connection().startSecureTransport();
+  }
+#if 0
+#endif
+
+  return Network::FilterStatus::Continue;
+}
+
+// Network::WriteFilter
+Network::FilterStatus StartTlsSwitchFilter::onWrite(Buffer::Instance&, bool) {
+  printf("Data Received onWrite!\n");
+  //data.drain(data.length());
```

Comment:
Do we need this?

Reply:
I think you are reviewing NOT the last version, as I removed those `printfs`.

Reply:
Oops... sorry by the noise!!!